File: app/csharp/AccountingSystem.Infrastructure/Persistence/Repositories/AuditLogRepository.cs
using AccountingSystem.Domain.Entities;
using AccountingSystem.Application.Ports.Out;
using AccountingSystem.Domain.Models.Audit;
using Dapper;
using Npgsql;
using System.Text.Json;

namespace AccountingSystem.Infrastructure.Persistence.Repositories;

/// <summary>
/// 監査ログリポジトリ実装
/// </summary>
public class AuditLogRepository : IAuditLogRepository
{
    private readonly string _connectionString;

    public AuditLogRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task<AuditLog> InsertAsync(AuditLog auditLog)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            INSERT INTO ""監査ログ"" (
                ""エンティティ種別"", ""エンティティID"", ""アクション"", ""ユーザーID"", ""ユーザー名"",
                ""タイムスタンプ"", ""変更前値"", ""変更後値"", ""変更内容"", ""理由"",
                ""IPアドレス"", ""ユーザーエージェント""
            ) VALUES (
                @EntityType, @EntityId, @Action, @UserId, @UserName,
                @Timestamp, @OldValues::jsonb, @NewValues::jsonb,
                @Changes::jsonb, @Reason, @IpAddress, @UserAgent
            )
            RETURNING ""監査ログID""";

        var id = await connection.ExecuteScalarAsync<long>(sql, new
        {
            auditLog.EntityType,
            auditLog.EntityId,
            Action = auditLog.Action.ToString(),
            auditLog.UserId,
            auditLog.UserName,
            auditLog.Timestamp,
            OldValues = SerializeToJson(auditLog.OldValues),
            NewValues = SerializeToJson(auditLog.NewValues),
            Changes = SerializeToJson(auditLog.Changes),
            auditLog.Reason,
            auditLog.IpAddress,
            auditLog.UserAgent
        });

        return auditLog with { Id = id };
    }

    public async Task<IReadOnlyList<AuditLog>> FindByEntityAsync(
        string entityType,
        string entityId)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT
                ""監査ログID"" AS Id,
                ""エンティティ種別"" AS EntityType,
                ""エンティティID"" AS EntityId,
                ""アクション"" AS Action,
                ""ユーザーID"" AS UserId,
                ""ユーザー名"" AS UserName,
                ""タイムスタンプ"" AS Timestamp,
                ""変更前値"" AS OldValues,
                ""変更後値"" AS NewValues,
                ""変更内容"" AS Changes,
                ""理由"" AS Reason,
                ""IPアドレス"" AS IpAddress,
                ""ユーザーエージェント"" AS UserAgent
            FROM ""監査ログ""
            WHERE ""エンティティ種別"" = @EntityType
              AND ""エンティティID"" = @EntityId
            ORDER BY ""タイムスタンプ"" DESC";

        var entities = await connection.QueryAsync<AuditLogEntity>(sql, new
        {
            EntityType = entityType,
            EntityId = entityId
        });

        return entities.Select(ToDomain).ToList();
    }

    public async Task<IReadOnlyList<AuditLog>> FindByUserAsync(
        string userId,
        DateTime startDate,
        DateTime endDate)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT
                ""監査ログID"" AS Id,
                ""エンティティ種別"" AS EntityType,
                ""エンティティID"" AS EntityId,
                ""アクション"" AS Action,
                ""ユーザーID"" AS UserId,
                ""ユーザー名"" AS UserName,
                ""タイムスタンプ"" AS Timestamp,
                ""変更前値"" AS OldValues,
                ""変更後値"" AS NewValues,
                ""変更内容"" AS Changes,
                ""理由"" AS Reason,
                ""IPアドレス"" AS IpAddress,
                ""ユーザーエージェント"" AS UserAgent
            FROM ""監査ログ""
            WHERE ""ユーザーID"" = @UserId
              AND ""タイムスタンプ"" BETWEEN @StartDate AND @EndDate
            ORDER BY ""タイムスタンプ"" DESC";

        var entities = await connection.QueryAsync<AuditLogEntity>(sql, new
        {
            UserId = userId,
            StartDate = startDate,
            EndDate = endDate
        });

        return entities.Select(ToDomain).ToList();
    }

    public async Task<IReadOnlyList<AuditLog>> FindByPeriodAsync(
        DateTime startDate,
        DateTime endDate,
        int limit)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT
                ""監査ログID"" AS Id,
                ""エンティティ種別"" AS EntityType,
                ""エンティティID"" AS EntityId,
                ""アクション"" AS Action,
                ""ユーザーID"" AS UserId,
                ""ユーザー名"" AS UserName,
                ""タイムスタンプ"" AS Timestamp,
                ""変更前値"" AS OldValues,
                ""変更後値"" AS NewValues,
                ""変更内容"" AS Changes,
                ""理由"" AS Reason,
                ""IPアドレス"" AS IpAddress,
                ""ユーザーエージェント"" AS UserAgent
            FROM ""監査ログ""
            WHERE ""タイムスタンプ"" BETWEEN @StartDate AND @EndDate
            ORDER BY ""タイムスタンプ"" DESC
            LIMIT @Limit";

        var entities = await connection.QueryAsync<AuditLogEntity>(sql, new
        {
            StartDate = startDate,
            EndDate = endDate,
            Limit = limit
        });

        return entities.Select(ToDomain).ToList();
    }

    public async Task<IReadOnlyList<AuditLog>> FindByActionAsync(
        AuditAction action,
        DateTime startDate,
        DateTime endDate,
        int limit)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT
                ""監査ログID"" AS Id,
                ""エンティティ種別"" AS EntityType,
                ""エンティティID"" AS EntityId,
                ""アクション"" AS Action,
                ""ユーザーID"" AS UserId,
                ""ユーザー名"" AS UserName,
                ""タイムスタンプ"" AS Timestamp,
                ""変更前値"" AS OldValues,
                ""変更後値"" AS NewValues,
                ""変更内容"" AS Changes,
                ""理由"" AS Reason,
                ""IPアドレス"" AS IpAddress,
                ""ユーザーエージェント"" AS UserAgent
            FROM ""監査ログ""
            WHERE ""アクション"" = @Action
              AND ""タイムスタンプ"" BETWEEN @StartDate AND @EndDate
            ORDER BY ""タイムスタンプ"" DESC
            LIMIT @Limit";

        var entities = await connection.QueryAsync<AuditLogEntity>(sql, new
        {
            Action = action.ToString(),
            StartDate = startDate,
            EndDate = endDate,
            Limit = limit
        });

        return entities.Select(ToDomain).ToList();
    }

    /// <summary>
    /// Entity → Domain Model 変換
    /// </summary>
    private static AuditLog ToDomain(AuditLogEntity entity)
    {
        return new AuditLog
        {
            Id = entity.Id,
            EntityType = entity.EntityType,
            EntityId = entity.EntityId,
            Action = Enum.Parse<AuditAction>(entity.Action),
            UserId = entity.UserId,
            UserName = entity.UserName,
            Timestamp = entity.Timestamp,
            OldValues = DeserializeFromJson(entity.OldValues),
            NewValues = DeserializeFromJson(entity.NewValues),
            Changes = DeserializeFromJson(entity.Changes),
            Reason = entity.Reason,
            IpAddress = entity.IpAddress,
            UserAgent = entity.UserAgent
        };
    }

    /// <summary>
    /// Dictionary を JSON 文字列にシリアライズ
    /// </summary>
    private static string? SerializeToJson(Dictionary<string, object>? dict)
    {
        return dict != null ? JsonSerializer.Serialize(dict) : null;
    }

    /// <summary>
    /// JSON 文字列を Dictionary にデシリアライズ
    /// </summary>
    private static Dictionary<string, object>? DeserializeFromJson(string? json)
    {
        return json != null
            ? JsonSerializer.Deserialize<Dictionary<string, object>>(json)
            : null;
    }

    /// <summary>
    /// Dapper マッピング用の Entity クラス
    /// </summary>
    private class AuditLogEntity
    {
        public long Id { get; set; }
        public string EntityType { get; set; } = string.Empty;
        public string EntityId { get; set; } = string.Empty;
        public string Action { get; set; } = string.Empty;
        public string UserId { get; set; } = string.Empty;
        public string UserName { get; set; } = string.Empty;
        public DateTime Timestamp { get; set; }
        public string? OldValues { get; set; }      // JSONB → String
        public string? NewValues { get; set; }      // JSONB → String
        public string? Changes { get; set; }        // JSONB → String
        public string? Reason { get; set; }
        public string? IpAddress { get; set; }
        public string? UserAgent { get; set; }
    }
}

File: app/fsharp/AccountingSystem.Infrastructure/Persistence/Repositories/AuditLogRepository.fs
namespace AccountingSystem.Infrastructure.Persistence.Repositories

open System
open System.Threading.Tasks
open Dapper
open Npgsql
open AccountingSystem.Domain.Types
open AccountingSystem.Domain.Models
open AccountingSystem.Infrastructure.Persistence.DAO

/// <summary>
/// 監査ログリポジトリ実装
/// </summary>
type AuditLogRepository(connection: NpgsqlConnection) =

    /// DAO → Domain Model 変換
    let toDomain (dao: AuditLogDao) : AuditLog option =
        AuditAction.fromCode dao.action
        |> Option.map (fun action ->
            {
                Id = Some dao.id
                EntityType = dao.entity_type
                EntityId = dao.entity_id
                Action = action
                UserId = dao.user_id
                UserName = dao.user_name
                Timestamp = dao.timestamp
                OldValues = if isNull dao.old_values then None else Some dao.old_values
                NewValues = if isNull dao.new_values then None else Some dao.new_values
                Changes = if isNull dao.changes then None else Some dao.changes
                Reason = if isNull dao.reason then None else Some dao.reason
                IpAddress = if isNull dao.ip_address then None else Some dao.ip_address
                UserAgent = if isNull dao.user_agent then None else Some dao.user_agent
            })

    /// <summary>
    /// 監査ログを保存する
    /// </summary>
    member _.InsertAsync(auditLog: AuditLog) : Task<AuditLog> =
        task {
            let sql = """
                INSERT INTO audit_log (
                    entity_type, entity_id, action, user_id, user_name,
                    timestamp, old_values, new_values, changes, reason,
                    ip_address, user_agent
                ) VALUES (
                    @EntityType, @EntityId, @Action, @UserId, @UserName,
                    @Timestamp, @OldValues::jsonb, @NewValues::jsonb,
                    @Changes::jsonb, @Reason, @IpAddress, @UserAgent
                )
                RETURNING id
            """

            let! id = connection.ExecuteScalarAsync<int64>(sql, {|
                EntityType = auditLog.EntityType
                EntityId = auditLog.EntityId
                Action = AuditAction.toCode auditLog.Action
                UserId = auditLog.UserId
                UserName = auditLog.UserName
                Timestamp = auditLog.Timestamp
                OldValues = auditLog.OldValues |> Option.toObj
                NewValues = auditLog.NewValues |> Option.toObj
                Changes = auditLog.Changes |> Option.toObj
                Reason = auditLog.Reason |> Option.toObj
                IpAddress = auditLog.IpAddress |> Option.toObj
                UserAgent = auditLog.UserAgent |> Option.toObj
            |})

            return { auditLog with Id = Some id }
        }

    /// <summary>
    /// エンティティ別の監査ログを取得する
    /// </summary>
    member _.FindByEntityAsync(entityType: string, entityId: string) : Task<AuditLog list> =
        task {
            let sql = """
                SELECT * FROM audit_log
                WHERE entity_type = @EntityType
                  AND entity_id = @EntityId
                ORDER BY timestamp DESC
            """

            let! daos = connection.QueryAsync<AuditLogDao>(sql, {|
                EntityType = entityType
                EntityId = entityId
            |})

            return
                daos
                |> Seq.choose toDomain
                |> Seq.toList
        }

    /// <summary>
    /// ユーザー別の監査ログを期間指定で取得する
    /// </summary>
    member _.FindByUserAsync(userId: string, startDate: DateTime, endDate: DateTime) : Task<AuditLog list> =
        task {
            let sql = """
                SELECT * FROM audit_log
                WHERE user_id = @UserId
                  AND timestamp BETWEEN @StartDate AND @EndDate
                ORDER BY timestamp DESC
            """

            let! daos = connection.QueryAsync<AuditLogDao>(sql, {|
                UserId = userId
                StartDate = startDate
                EndDate = endDate
            |})

            return
                daos
                |> Seq.choose toDomain
                |> Seq.toList
        }

    /// <summary>
    /// 期間内のすべての監査ログを取得する
    /// </summary>
    member _.FindByDateRangeAsync(startDate: DateTime, endDate: DateTime) : Task<AuditLog list> =
        task {
            let sql = """
                SELECT * FROM audit_log
                WHERE timestamp BETWEEN @StartDate AND @EndDate
                ORDER BY timestamp DESC
            """

            let! daos = connection.QueryAsync<AuditLogDao>(sql, {|
                StartDate = startDate
                EndDate = endDate
            |})

            return
                daos
                |> Seq.choose toDomain
                |> Seq.toList
        }

    /// <summary>
    /// アクション種別で監査ログを検索する
    /// </summary>
    member _.FindByActionAsync(action: string, startDate: DateTime, endDate: DateTime) : Task<AuditLog list> =
        task {
            let sql = """
                SELECT * FROM audit_log
                WHERE action = @Action
                  AND timestamp BETWEEN @StartDate AND @EndDate
                ORDER BY timestamp DESC
            """

            let! daos = connection.QueryAsync<AuditLogDao>(sql, {|
                Action = action
                StartDate = startDate
                EndDate = endDate
            |})

            return
                daos
                |> Seq.choose toDomain
                |> Seq.toList
        }
